### DISCUSSION TITLE
Error when running queues

### DISCUSSION THREAD

**Discussion by sklr** [laravel/framework#51079] [Q&A]
### Laravel Version

11.3.1

### PHP Version

8.3.6

### Description

When I run a queue with php artisan queue:work no error is displayed in CLI, but in storage/logs/laravel.log there is multiple:
```
[2024-04-15 13:33:19] local.ERROR: abs(): Argument #1 ($num) must be of type int|float, string given {"exception":"[object] (TypeError(code: 0): abs(): Argument #1 ($num) must be of type int|float, string given at /var/www/example.com/vendor/nesbot/carbon/src/Carbon/Traits/Units.php:162)
[stacktrace]
#0 /var/www/example.com/vendor/nesbot/carbon/src/Carbon/Traits/Units.php(162): abs()
#1 /var/www/example.com/vendor/nesbot/carbon/src/Carbon/Traits/Units.php(49): Carbon\\Carbon->addUTCUnit()
#2 /var/www/example.com/vendor/nesbot/carbon/src/Carbon/Traits/Date.php(2662): Carbon\\Carbon->addRealUnit()
#3 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Support/InteractsWithTime.php(38): Carbon\\Carbon->__call()
#4 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/RedisQueue.php(278): Illuminate\\Queue\\Queue->availableAt()
#5 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/RedisQueue.php(228): Illuminate\\Queue\\RedisQueue->retrieveNextJob()
#6 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(349): Illuminate\\Queue\\RedisQueue->pop()
#7 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(363): Illuminate\\Queue\\Worker->Illuminate\\Queue\\{closure}()
#8 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/Worker.php(162): Illuminate\\Queue\\Worker->getNextJob()
#9 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(139): Illuminate\\Queue\\Worker->daemon()
#10 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php(122): Illuminate\\Queue\\Console\\WorkCommand->runWorker()
#11 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): Illuminate\\Queue\\Console\\WorkCommand->handle()
#12 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Container/Util.php(41): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#13 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#14 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(35): Illuminate\\Container\\BoundMethod::callBoundMethod()
#15 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Container/Container.php(662): Illuminate\\Container\\BoundMethod::call()
#16 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Console/Command.php(212): Illuminate\\Container\\Container->call()
#17 /var/www/example.com/vendor/symfony/console/Command/Command.php(279): Illuminate\\Console\\Command->execute()
#18 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Console/Command.php(181): Symfony\\Component\\Console\\Command\\Command->run()
#19 /var/www/example.com/vendor/symfony/console/Application.php(1049): Illuminate\\Console\\Command->run()
#20 /var/www/example.com/vendor/symfony/console/Application.php(318): Symfony\\Component\\Console\\Application->doRunCommand()
#21 /var/www/example.com/vendor/symfony/console/Application.php(169): Symfony\\Component\\Console\\Application->doRun()
#22 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php(196): Symfony\\Component\\Console\\Application->run()
#23 /var/www/example.com/artisan(35): Illuminate\\Foundation\\Console\\Kernel->handle()
#24 {main}
"} 
```

I think this is due to Carbon using `declare(strict_types=1);` in `carbon/src/Carbon/Traits/Units.php`. When `$unit = "second"`, the $value is never cast/converted to float|int.

Carbon version is 3.2.4

### Steps To Reproduce

Just run a queue worker. No jobs are actually queued.

**Comment by sxtnmedia** [comment]
Fresh Laravel v11.3.1
Carbon 3.2.4
PHP 8.3.6

Couldn't reproduce the error.

**Comment by devajmeireles** [comment]
Same here, @sxtnmedia 👍🏻

**Comment by crynobone** [comment]
Hey there, thanks for reporting this issue.

We'll need more info and/or code to debug this further. Can you please create a repository with the command below, commit the code that reproduces the issue *as one separate commit* on the main/master branch and share the repository here?

Please make sure that you have the [latest version of the Laravel installer](https://github.com/laravel/installer) in order to run this command. Please also make sure you have both Git & [the GitHub CLI tool](https://cli.github.com/) properly set up.

    laravel new bug-report --github="--public"

Do not amend and create a separate commit with your custom changes. After you've posted the repository, we'll try to reproduce the issue.

Thanks!

**Comment by sklr** [comment]
I forgot to mention that this happened after an upgrade from Laravel 9 to 10 to 11. I will try to reproduce the issue in a separate repo.

**Comment by pakistanimoon** [comment]
Could not reproduce Error on Fresh Laravel
Can you please show some code snippets or upload repo?

**Comment by khamazaspyan** [comment]
I have same issue in Laravel 11, php 8.3.4, carbon 3.2.4.
How to fix it?

**Reply by ghost** [reply]
The latest version of laravel 11 has here
#3 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Support/InteractsWithTime.php(38): Carbon\\Carbon->__call()

https://github.com/laravel/framework/blob/1ab4de52d0ac3a7a48410bf1e7b735499ce59284/src/Illuminate/Support/InteractsWithTime.php#L26C1-L39C6


``` 
   /**
     * Get the "available at" UNIX timestamp.
     *
     * @param  \DateTimeInterface|\DateInterval|int  $delay
     * @return int
     */
    protected function availableAt($delay = 0)
    {
        $delay = $this->parseDateInterval($delay);

        return $delay instanceof DateTimeInterface
                            ? $delay->getTimestamp()
                            : Carbon::now()->addRealSeconds($delay)->getTimestamp();
    }
```

I guess that $delay is int and it might be called from #4 /var/www/example.com/vendor/laravel/framework/src/Illuminate/Queue/RedisQueue.php(278): Illuminate\\Queue\\Queue->availableAt()
https://github.com/laravel/framework/blob/1ab4de52d0ac3a7a48410bf1e7b735499ce59284/src/Illuminate/Queue/RedisQueue.php#L278

it seems like this https://github.com/briannesbitt/Carbon/blob/3a103ffea3709795029166da901a23fbc781db14/src/Carbon/Traits/Date.php#L2608

is called from https://github.com/laravel/framework/blob/1ab4de52d0ac3a7a48410bf1e7b735499ce59284/src/Illuminate/Support/InteractsWithTime.php#L38C46-L38C60

Maybe in between the $parameters are changing...


 * @method        $this               addSeconds(int $value = 1)                                                                      Add seconds (the $value count passed in) to the instance (using date interval).
 vs
 * @method        $this               addRealSeconds(int $value = 1)                                                                  Add seconds (the $value count passed in) to the instance (using timestamp).

... or 
```
        return new RedisQueue(
            $this->redis, $config['queue'],
            $config['connection'] ?? $this->connection,
            $config['retry_after'] ?? 60,
            $config['block_for'] ?? null,
            $config['after_commit'] ?? null
        );
```

OR the RedisQueue is instantiated with string in its $config['retry_after']

but...
https://github.com/laravel/framework/blob/1ab4de52d0ac3a7a48410bf1e7b735499ce59284/config/queue.php#L70

Conclusion pls check queue.php config for redis.

**Comment by sk33wiff** [comment]
I've been having the same issue as well after upgrading from Laravel 10 to 11.
It's happening when dispatching a `ShouldBeUniqueUntilProcessing` job that uses DB for locks.

@khamazaspyan, @sklr did you managed to solve this issue?

UPDATE: in my case I can confirm that the exception is being thrown from [Carbon/Traits/Units.php#L162](https://github.com/briannesbitt/Carbon/blob/3.8.2/src/Carbon/Traits/Units.php#L162)
Looks like some part of the Laravel code is passing $value to `Units::addUTCUnit` as a string instead of an int.
